README
= PGQ-PHP

== Introduction

This project provides PHP level API to use the PGQ SQL API.

Read more about PGQ here:
  http://skytools.projects.postgresql.org/doc/pgq-sql.html[]
  http://www.pgcon.org/2008/schedule/events/79.en.html[]

It then provides several kinds of Consumer that you'll want to
extends. All the consumer wants you to implement your own
process_event(&$event) callback function, and will care about PGQ
inner fonctionning.

PGQ is about producing events in a (work) queue and having one or more
consumer get those events, in batches, and work on them. What the PHP
class to be presented here do is fetching the batches and calling your
own process_event() function for each event.

== Tools

=== SimpleLogger

This class handles logging with a +loglevel+: only messages more
important than current +loglevel+ are written to the logfile. The
levels are, by increasing order of importance:

* DEBUG
* VERBOSE
* NOTICE
* WARNING
* ERROR
* FATAL

The +SimpleLogger+ offers a method for each +loglevel+, taking a
format string then arguments, just like +sprintf()+ does. You have for
example 

  $logger->debug("a message with a '%s' string", $str);

=== SystemDaemon

This class implements a System Daemon controlled with commands. The
principle is to run in the background and loop forever, taking a
+$this->delay+ rest whenever possible between two loops.

+SystemDaemon+ uses SimpleLogger in +$this->log+ instance, and have it
reopens its logfile at +reload+ time. This means it plays well with
logrotate, just don't forget to +reload+ the daemon at logrotate
postprocess time.

The commands available for any +SystemDaemon+ are: +start+, +stop+,
+kill+, +reload+, +restart+, +status+, +logless+, +logmore+.

The difference between +stop+ and +kill+ is that in the first case the
daemon will terminate current processing and only stops at next
opportunity to sleep for +$this->delay+, while +kill+ forces it into
exiting (calling +$this->stop()+) first.

The +reload+ command will have the daemon call user defined
+$this->reload()+ command at next sleep opportunity. This command
could e.g. read a configuration file and change settings.

Commands +logless+ and +logmore+ are considered without delay and
makes the used logger to get instantly more or less verbose by
increasing or decreasing its +loglevel+. This is useful when you want
to know exactly what a daemon is doing now, but can't afford to
restart it.

When implementing a +SystemDaemon+, you get to just implement
+$this->config()+ and +$this->process()+ functions, which will get
called in the infinite looping. The former will only get called when a
reload has been ordered (+SIGHUP+), and the latter in each and every
loop.

The +SystemDaemon+ will also register its +SimpleLogger+ instance as
the PHP error handler, and consider quitting when confronted to a PHP
FATAL errors (one of +E_STRICT+, +E_PARSE+, +E_CORE_ERROR+,
+E_CORE_WARNING+, +E_COMPILE_ERROR+, +E_COMPILE_WARNING+), and will
call user defined +$this->php_error_hook()+ for PHP +E_ERROR+ or
+E_USER_ERROR+ level errors.

=== PGQ

This class is abstract and has a +public static+ method for each SQL
function that the PGQ SQL API offers. If PHP knew about namespaces or
modules, this would be a module.

== Consumers

Those class are the one handling how PGQ really works. They will get
batches from the queue, get events from them, allow you to process the
event by calling user defined +$this->process_event(&$event)+
function, and call finish_batch() as appropriate.

The +$this->process_event()+ function should return one of those
return codes:

PGQ_EVENT_OK::

  When the event processing is satisfactory.

PGQ_EVENT_FAILED::

  The event is tagged as failed with +$event->failed_reason+ reason,
  and when using +PGQEventRemoteConsumer+ the related processing done
  on remote database is rollbacked.

PGQ_EVENT_RETRY::

  The event is tagged as retry and will get reinserted into main queue
  after a minimum delay of +$event->retry_delay+ seconds. When using
  +PGQEventRemoteConsumer+ the related processing done on the remote
  database is rollbacked.

PGQ_ABORT_BATCH::

  All current batch processing is canceled, at both remote and queue
  database when using some sort of +RemoteConsumer+. The batch is not
  finished, so you'll get the events back at next run(s).

=== PGQConsumer

This is a +SystemDaemon+ which implements some more commands in order
to install the queue and register the consumer.

When extending a +PGQConsumer+, you get to implement +$this->config()+
and +$this->process_event()+ functions, and you're done.

The constructor needs a queue name (+qname+), a consumer name
(+cname+) and a database connection string.

The added commands are:

install::
  Create the queue +qname+ and register a consumer +cname+.

uninstall::
  Unregister the consumer +cname+ and drop the queue +qname+.

check::
  True only if the queue +qname+ exists and the consumer +cname+ is registered.

create_queue::
  Create the queue +qname+.

drop_queue::
  Drop the queue +qname+.

register::
  Register the consumer +cname+.

unregister::
  Unregister the consumer +cname+.

failed::
  Print out a list of failed events for queue +qname+ and consumer
  +cname+.

delete::
  Delete given event id, or all failed events if given +all+ as an
  event id.

retry::
  Retry given event id, or all failed events if given +all+ as an
  event id.

=== PGQInteractiveConsumer

This class assume the looping will get done elsewhere, at the calling
site for example. It consumes all available events (up until
next_batch() returns +null+).

The lag is not controlled by the implementer class but rather by the
user of it.

Implementer have to call +$this->process()+, which will start
consuming all available events and call the
+$this->process_event(&$event)+ hook for each event.

Internal design note::
  
  +PGQInteractiveRemoteConsumer+ needs to implement all PGQ methods
  for itself because of PHP limitation of extending from only one base
  class: +PGQConsumer+ could not extends both +SystemDaemon+ and
  +PGQClass+, where we should put the class abstraction over the API
  module.

=== PGQRemoteConsumer

+PGQRemoteConsumer+ is a +PGQConsumer+ controlling two PostgreSQL
connections and which will handle +COMMIT+ and +ROLLBACK+ nicely on
both of them.

This means you want to use +PGQRemoteConsumer+ when you're processing
events from one database and apply changes to another one, the remote
one. The +PGQRemoteConsumer+ takes advantage of the fact that the
remote processing is transactionnal (happens on a database) to get
sure any +COMMIT+ ed work on remote connection is associated with
events properly consumed.

Any error in event consuming or remote processing will cause the
current batch processing to be +ROLLBACK+ ed at both points, meaning
the events will get consumed again later.

=== PGQEventRemoteConsumer

When you need to be able to +COMMIT+ or +ROLLBACK+ both transaction at
event level, +PGQEventRemoteConsumer+ is what you're after. It will
use a subtransaction (+SAVEPOINT+) for each event and will be able to
+ROLLBACK TO SAVEPOINT+ on the remote side for any processing error
related to a single event processing.

=== PGQCoopeConsumer

This Consumer will share batches in between all its subconsumer processes.
You need to register each of the subconsumer separately.


debian/changelog
libphp-pgq (0.15-2) unstable; urgency=low

  * Fix Cooperative Consumer status

 -- Dimitri Fontaine <dim@tapoueh.org>  Wed, 18 Apr 2012 11:42:41 +0200

libphp-pgq (0.14-1) unstable; urgency=low

  * Add support for PGQ Cooperative Consumers (PGQ 3)

 -- Dimitri Fontaine <dim@tapoueh.org>  Wed, 08 Feb 2012 15:35:40 +0100

libphp-pgq (0.13-1) unstable; urgency=low

  * Fix simple logger streams detection

 -- Dimitri Fontaine <dim@tapoueh.org>  Thu, 22 Dec 2011 12:17:26 +0100

libphp-pgq (0.12-1) unstable; urgency=low

  * Better DEBUG facilities for unhandled errors
  * Add handling of errors in newer versions of PHP

 -- Dimitri Fontaine <dim@tapoueh.org>  Thu, 01 Dec 2011 14:43:51 +0100

libphp-pgq (0.11-1) unstable; urgency=low

  * E_STRICT is not a FATAL error by any means (try being backward compatible)

 -- Dimitri Fontaine <dim@tapoueh.org>  Mon, 31 Oct 2011 10:42:13 +0100

libphp-pgq (0.10-1) unstable; urgency=low

  * split() is deprecated, use explode()
  * fix a WARNING about already closed STDOUT

 -- Dimitri Fontaine <dim@tapoueh.org>  Tue, 04 Oct 2011 10:47:42 +0200

libphp-pgq (0.9~dev1-2) unstable; urgency=low

  * Fix the urlencoded decode function to avoid using empty()

 -- Dimitri Fontaine <dim@tapoueh.org>  Mon, 08 Feb 2010 10:14:09 +0100

libphp-pgq (0.8-1) unstable; urgency=low

  * Avoid defining constants if they are already set, by our users e.g.
  * Fix exception handling in function exceptFault()
  * As tempnam() creates the file, append .log by means of rename

 -- Dimitri Fontaine <dim@tapoueh.org>  Tue, 01 Dec 2009 16:39:27 +0100

libphp-pgq (0.7-1) unstable; urgency=low

  * add a kill_hook($pid) in SystemDaemon
  * Fix the fatal error hint message to create the pgq_last_batch table

 -- Dimitri Fontaine <dim@tapoueh.org>  Thu, 19 Nov 2009 17:51:41 +0100

libphp-pgq (0.6-1) unstable; urgency=low

  * Protect install / uninstall in a transaction
  * Debug PGQ::get_consumers() to only use $result when not False
  * Debug PGQ::{un}register to preperly define result variables...

 -- Dimitri Fontaine <dim@tapoueh.org>  Tue, 15 Sep 2009 10:33:58 +0200

libphp-pgq (0.5-1) unstable; urgency=low

  * Add pgq.maint_retry_events() in the PGQ and PGQConsumer APIs
  * change the default PGQEvent retry_delay to 5 hours instead of null

 -- Dimitri Fontaine <dim@tapoueh.org>  Thu, 25 Jun 2009 10:36:50 +0200

libphp-pgq (0.4-1) unstable; urgency=low

  * Fix failed_event_list() to fill an associative array of event->id
  * Fix print Array() from failed_event_{delete,retry}
  * Connect() errors are now transient, no more fatal
  * rename deconnect() to disconnect()

 -- Dimitri Fontaine <dim@tapoueh.org>  Wed, 12 Nov 2008 12:50:30 +0100

libphp-pgq (0.3-1) unstable; urgency=low

  * Do not unregister consumer when there's a data loss risk identified

 -- Dimitri Fontaine <dim@tapoueh.org>  Thu, 11 Sep 2008 16:09:37 +0200

libphp-pgq (0.2-1) unstable; urgency=low

  * FIX require path to include "pgq/", as implied by the packaging
  * FIX some typos and add PGQ.php

 -- Dimitri Fontaine <dim@tapoueh.org>  Wed, 25 Jun 2008 09:25:36 +0200

libphp-pgq (0.1-1) unstable; urgency=low

  * Initial debian packaging

 -- Dimitri Fontaine <dim@tapoueh.org>  Tue, 24 Jun 2008 11:15:49 +0200



PGQConsumer.php
<?php
require_once("pgq/PGQ.php");
require_once("pgq/SystemDaemon.php");

/**
 * PGQConsumer is a SystemDaemon providing the PGQ SQL API for PHP
 * applications, and implementing a simple Consumer model atop of it.
 *
 * Depending on your needs, please check more specialized classes
 * PGQRemoteConsumer and PGQEventRemoteConsumer.
 *
 * Online references about PGQ are to be found here:
 *  http://skytools.projects.postgresql.org/doc/pgq-admin.html
 *  http://skytools.projects.postgresql.org/doc/pgq-sql.html
 *  http://skytools.projects.postgresql.org/doc/pgq-nodupes.html
 *
 *  http://kaiv.wordpress.com/2007/10/19/skytools-database-scripting-framework-pgq/
 *
 *  http://www.pgcon.org/2008/schedule/events/79.en.html
 *
 * When implementing a PGQConsumer, you should define following methods:
 *  - php_error_hook()
 *  - config()
 *  - process_event(&$event)  --- $event is a PGQEvent
 *
 * See SystemDaemon header documentation for details about first two
 * functions.
 *
 * The consumer model provided by PGQConsumer works by letting the
 * inheriting worker define how to process a single event. The
 * function process_event() will get called while in a transaction, so
 * don't issue COMMIT or ROLLBACK on $this->pg_src_con.
 *
 * You should return one of those constants from process_event():
 * PGQ_EVENT_OK, PGQ_EVENT_FAILED, PGQ_EVENT_RETRY or PGQ_ABORT_BATCH.
 *
 * The last will force ROLLBACK, use it when you encounter PostgreSQL
 * level errors: the transactions are no more valid, you return
 * PGQ_ABORT_BATCH and PGQConsumer will handle: batch is not finished,
 * the error is considered not fatal.
 *
 * When you return PGQ_EVENT_RETRY or PGQ_EVENT_FAILED, PGQConsumer
 * will call the matching pgq SQL function to tag the Event. In the
 * retry case, PGQConsumer will get $event->retry_delay attribute to
 * give to the SQL level API, and in the failed case,
 * $event->failed_reason.
 *
 * PGQConsumer supports same commands as SystemDaemon, and some more:
 *   install, uninstall, create_queue, drop_queue, register, unregister
 *   failed, delete <event_id|all>, retry <event_id|all>
 *
 * See SystemDaemon documentation for using this class at PHP and
 * command line levels.
 */

abstract class PGQConsumer extends SystemDaemon
{
  protected $connected  = False;
  protected $src_constr;
  protected $pg_src_con;

  protected $cname;  // consumer name (pgq consumer id)
  protected $qname; // queue name

  public function __construct($cname, $qname, $argc, $argv, $src_constr)
  {
    $this->cname = $cname;
    $this->qname = $qname;

    $this->commands[] = "install";
    $this->commands[] = "uninstall";
    $this->commands[] = "check";
    $this->commands[] = "create_queue";
    $this->commands[] = "drop_queue";
    $this->commands[] = "register";
    $this->commands[] = "unregister";
    $this->commands[] = "failed";
    $this->commands[] = "delete";
    $this->commands[] = "retry";

    $this->src_constr  = $src_constr;

    parent::__construct($argc, $argv);
  }

  /**
   * Implement those functions when inheriting from this class.
   */
  public function config() {}

  /**
   * Optionnal hooks
   */
  public function preprocess_event($event) {}
  public function postprocess_event($event) {}

  /**
   * We overload SystemDaemon::main() in order to support our specific
   * commands too.
   */
  public function main($argc, $argv) {
    if( $argc < 2 ) {
      fprintf(STDERR, $this->usage($this->name));
      exit(1);
    }

    switch( $argv[1] )
      {
      case "install":
	$pid = $this->getpid();
	if( $pid !== False )
	  $this->log->fatal("Can not install an already running daemon.");
	else {
	  $ret = $this->install();
	  exit($ret ? 0 : 1);
	}
	break;

      case "uninstall":
	$pid = $this->getpid();
	if( $pid !== False )
	  $this->log->fatal("Can not uninstall a running daemon.");
	else {
	  $ret = $this->uninstall();
	  exit($ret ? 0 : 1);
	}
	break;

      case "check":
	$pid = $this->getpid();
	if( $pid !== False ) {
	  $this->log->fatal("Daemon already running.");
	  $this->status();
	}
	else {
	  $ret = $this->check();
	  exit($ret ? 0 : 1);
	}
	break;

      case "create_queue":
	$pid = $this->getpid();
	if( $pid !== False )
	  $this->log->fatal("Can not create queue for an already running daemon.");
	else {
	  if( $this->connect() === False ) {
	    exit(3);
	  }
	  else {
	    $ret = $this->create_queue();
	    $this->disconnect();
	    exit($ret ? 0 : 1);
	  }
	}
	break;

      case "drop_queue":
	$pid = $this->getpid();
	if( $pid !== False )
	  $this->log->fatal("Can not drop queue for a running daemon.");
	else {
	  if( $this->connect() === False ) {
	    exit(3);
	  }
	  else {
	    $ret = $this->drop_queue();
	    $this->disconnect();
	    exit($ret ? 0 : 1);
	  }
	}
	break;

      case "register":
	$pid = $this->getpid();
	if( $pid !== False )
	  $this->log->fatal("Can not register already running daemon.");
	else {
	  if( $this->connect() === False ) {
	    exit(3);
	  }
	  else {
	    $ret = $this->register();
	    $this->disconnect();

	    exit($ret ? 0 : 1);
	  }
	}
	break;

      case "unregister":
	$pid = $this->getpid();
	if( $pid !== False )
	  $this->log->fatal("Can not unregister already running daemon.");
	else {
	  $force = False;
	  if( $argc >= 3 ) {
	    $force = ($argv[2] == "--force");

	    if( $force != True ) {
	      $this->log->fatal("unregister [--force]");
	      exit(1);
	    }
	  }

	  if( $this->connect() === False ) {
	    exit(3);
	  }
	  else {
	    if( $force )
	      $ret = $this->force_unregister();
	    else {
	      if( $this->check_unregister() )
		$ret = $this->force_unregister();
	      else
		$ret = False;
	    }

	    $this->disconnect();

	    exit($ret ? 0 : 1);
	  }
	}
	break;

      case "failed":
	if( $this->connect() === False ) {
	  exit(3);
	}
	else {
	  $events = $this->failed_event_list();
	  if( $events !== False )
	    foreach( $events  as $event ) {
	      echo $event."\n";
	    }
	  else
	    $this->log->warning("Failed event list is empty.");
	  $this->disconnect();
	  exit(0);
	}
	break;

      case "delete":
	if( $argc < 3 ) {
	  $this->log->fatal("delete <event_id> [<event_id> ...]");
	  exit(1);
	}

	if( $this->connect() === False ) {
	  exit(3);
	}
	else {
	  if( $argv[2] == "all" )
	    $this->failed_event_delete_all();
	  else {
	    for($i = 2; $i < $argc; $i++) {
	      $this->failed_event_delete($argv[$i]);
	    }
	  }
	  $this->disconnect();
	  exit(0);
	}
	break;

      case "retry":
	if( $argc < 3 ) {
	  $this->log->fatal("delete <event_id> [<event_id> ...]");
	  exit(1);
	}
	if( $this->connect() === False ) {
	  exit(3);
	}
	else {
	  if( $argv[2] == "all" )
	    $this->failed_event_retry_all();
	  else {
	    for($i = 2; $i < $argc; $i++) {
	      $this->failed_event_retry($argv[$i]);
	    }
	  }
	  $this->disconnect();
	  exit(0);
	}
	break;

      default:
	/**
	 * Support daemon commands: start, stop, reload, restart, ...
	 */
	parent::main($argc, $argv);
	break;
      }
  }

  /**
   * Check installation is ok before to run.
   */
  public function run() {
    if( $this->check() )
      parent::run();
    else
      $this->stop();
  }

  /**
   * Stop is called either at normal exit or when receiving an error,
   * PHP errors included.
   */
  public function stop() {
    if( $this->connected ) {
      $this->rollback();
      $this->disconnect();
    }
    parent::stop();
  }

  /**
   * Process available batches, sleeping only when next_batch() returns null.
   *
   * We want to consume empty batches as fast as possible to avoid
   * lagging behind.
   *
   * Sleep only when next_batch() is null.
   *  null:  no more batches awaiting us to process, we sleep
   *  True:  successfully processed a batch, have it processed
   *  False: failed to get next_batch, sleep (temp. failure?)
   *
   * The sleeping is cared about at SystemDaemon level, in run()
   * method, between process() calls.
   */
  public function process()
  {
    $sleep = False;

    if( $this->connect() === False ) {
      // Can't connect to database, transient error: return to sleep
      return;
    }

    while( ! $sleep && ! $this->killed ) {
      $batch_id = $this->next_batch();

      switch( $batch_id ) {
      case null:
	$this->log->verbose("PGQConsumer.process: next_batch is null, sleep.");
	$sleep = True;
	break;

      case False:
	$this->log->verbose("PGQConsumer.process: failed to get batch.");
	break;

      case True:
	$this->log->verbose("PGQConsumer.process: \$this->process_batch(%d).",
			    $batch_id);

	$this->process_batch($batch_id);
	break;
      }
    }
    $this->disconnect();
  }

  /**
   * We want that all events of a same batch to have been either
   * processed or postponed by the end of this loop, so we only
   * COMMIT; when no error arise.
   */
  public function process_batch($batch_id) {
    $events = $this->preprocess_batch($batch_id);

    if( $events === False) {
	$this->log->verbose("PGQConsumer.preprocess_batch got not events (False).");
      return False;
    }

    /**
     * Event processing loop!
     */
    $abort_batch = False;

    foreach( $events as $event ) {
      if( $abort_batch )
	break;

      if( $this->preprocess_event($event) === PGQ_ABORT_BATCH ) {
	$abort_batch = True;
      }
      else {
	$tag = $event->tag( $this->process_event($event) );
	$this->log->verbose("PGQConsumer.process_batch processed event %d of batch %d, tag %d",
	                    $event->id, $batch_id, $tag);

	switch( $tag  ) {
	case PGQ_ABORT_BATCH:
	  $this->log->verbose("PGQConsumer.process_batch got PGQ_ABORT_BATCH");
	  $abort_batch = True;
	  break;

	case PGQ_EVENT_FAILED:
	  $this->event_failed($batch_id, $event);
	  break;

	case PGQ_EVENT_RETRY:
	  $this->event_retry($batch_id, $event);
	  break;

	case PGQ_EVENT_OK:
	  break;
	}

	if( ! $abort_batch ) {
	  if( $this->postprocess_event($event) === PGQ_ABORT_BATCH) {
	    $this->log->verbose("PGQConsumer.postprocess_event(%d) abort batches", $event->id);
	    $abort_batch = True;
	  }
	}
      }
    }
    return $this->postprocess_batch($batch_id, $abort_batch);
  }

  /**
   * Prepare batch processing, returns PGQEvent list or False if
   * empty.
   *
   * Exists as a separate function for implementers to accomodate
   * easily, see PGQRemoteConsumer.
   */
  public function preprocess_batch($batch_id) {
    pg_query($this->pg_src_con, "BEGIN;");

    $events = $this->get_batch_events($batch_id);

    if( $events === False ) {
      // batch with no event
      $this->log->debug("PGQConsumer.preprocess_batch got no events");
      $this->finish_batch($batch_id);
      pg_query($this->pg_src_con, "COMMIT;");
      return False;
    }
    $this->log->debug("PGQConsumer.preprocess_batch got %d events", count($events));
    return $events;
  }

  /**
   * Conclude batch processing, COMMIT or ROLLBACK if $abort_batch,
   * and call finish_batch().
   */
  public function postprocess_batch($batch_id, $abort_batch = False) {
    if( $abort_batch) {
    	$this->log->warning("PGQConsumer.postprocess_batch aborts: ROLLBACK");
      $this->rollback();
      return False;
    }

    if( $this->finish_batch($batch_id) === False )
      $this->log->warning("Could not mark batch id %d as finished", $batch_id);

    pg_query($this->pg_src_con, "COMMIT;");
    return True;
  }

  /**
   * print out deamon status, and detailed queue informations
   */
  public function status() {
    $pid = $this->getpid();
    if( $pid === False )
      printf("PGQConsumer %s is not running.\n", $this->name);
    else {
      printf("PGQConsumer %s is running with pid %d\n", $this->name, $pid);

      if( $this->connect() === False ) {
	return;
      }
      $status = $this->get_consumer_info();
      foreach( $status as $k => $v)
	printf("%s: %s\n", $k, $v);

      $this->disconnect();
    }
  }

  /**
   * Consumer installation: create the queue and register consumer.
   */
  protected function install() {
    if( $this->connect() === False )
      return False;

    pg_query($this->pg_src_con, "BEGIN;");
    $ret = $this->create_queue();

    if( $ret ) {
      $ret = $this->register();
    }
    pg_query($this->pg_src_con, "COMMIT;");

    $this->disconnect();
    return $ret;
  }

  /**
   * Consumer uninstall: unregister the consumer and drop the queue
   */
  protected function uninstall() {
    if( $this->connect() === False )
      return False;

    pg_query($this->pg_src_con, "BEGIN;");
    $ret = $this->unregister();

    if( $ret ) {
      $ret = $this->drop_queue();
    }
    pg_query($this->pg_src_con, "COMMIT;");

    $this->disconnect();
    return $ret;
  }

  /**
   * Consumer check: check the queue exists and the consumer is
   * registered.
   */
  protected function check() {
    if( $this->connect() === False )
      return False;

    $ret = $this->queue_exists();

    if( $ret ) {
      $ret = $this->is_registered();
    }
    $this->disconnect();

    return $ret;
  }

  /**
   * check_unregister returns True when we can unregister without
   * loosing data.
   *
   * This now means there's some other consumer registered.
   * TODO: check for trigger absence when only 1 consumer is registered (us).
   */
  protected function check_unregister() {
    $consumers = $this->get_consumers();

    if( $consumers === False ) {
      $this->log->error("Couldn't get consumer list for %s", $this->qname);
      return False;
    }

    if ( count($consumers) > 1 ) {
      $this->log->notice("Some other consumers are registered, ".
			 "unregistering %s.", $this->cname);
      return True;
    }
    else {
      // We are the only one consumer
      $this->log->warning("This %s consumer is the only one registered, ".
			  "use --force to unregister it.", $this->cname);
      return False;
    }
    return False;
  }

  /**
   * unregister without bothering.
   */
  protected function force_unregister() {
    return $this->unregister();
  }

  /**
   * TODO.
   * drop triggers on the provider
   */
  protected function drop_trigger() {
    return False;
  }

  /**
   * TODO.
   *
   * keep data job is to go find non consumed events in the pgq queue
   * and insert them again in the main table.
   */
  protected function keep_data() {
    if( $this->drop_trigger() ) {
      return False;
    }
    return False;
  }

  /**
   * Connects to the conw & conp connection strings.
   */
  public function connect($force = False) {
    if( $this->connected && ! $force ) {
      $this->log->notice("connect called when connected is True");
      return True;
    }

    if( $this->src_constr != "" ) {
      $this->log->verbose("Opening pg_src connexion '%s'.", $this->src_constr);
      $this->pg_src_con = pg_connect($this->src_constr);

      if( $this->pg_src_con === False ) {
	$this->log->error("Could not open pg_src connection '%s'.",
			  $this->src_constr);
	// $this->stop();
	return False;
      }
    }
    else {
      $this->log->error("Connexion string is empty, please fix.");
      return False;
    }

    $this->connected = True;
    return True;
  }

  /**
   * Disconnect from databases
   */
  public function disconnect() {
    if( ! $this->connected ) {
      $this->log->notice("disconnect called when $this->connected is False");
      return;
    }

    if( $this->pg_src_con != null && $this->pg_src_con !== False ) {
      $this->log->verbose("Closing pg_src connection '%s'.",
			  $this->src_constr);
      pg_close($this->pg_src_con);
      $this->pg_src_con = null;
    }
    $this->connected = False;
  }

  /**
   * ROLLBACK ongoing transactions on src & dst connections
   */
  protected function rollback() {
    if( $this->pg_src_con != null && $this->pg_src_con !== False ) {
      $this->log->notice("ROLLBACK pg_src connection '%s'.",
			 $this->src_constr);
      pg_query($this->pg_src_con, "ROLLBACK;");
    }
  }

  /**
   * This hook is called when a PHP error at level
   * E_USER_ERROR or E_ERROR is raised.
   *
   * Those errors are not considered fatal: abort current batch
   * processing, but don't have the PGQConsumer daemon quit.
   */
  protected function php_error_hook() {
    $this->rollback();
  }


  protected function create_queue() {
    return PGQ::create_queue($this->log, $this->pg_src_con, $this->qname);
  }

  protected function drop_queue() {
    return PGQ::drop_queue($this->log, $this->pg_src_con, $this->qname);
  }

  protected function queue_exists() {
    return PGQ::queue_exists($this->log, $this->pg_src_con, $this->qname);
  }

  protected function register() {
    return PGQ::register($this->log, $this->pg_src_con,
			 $this->qname, $this->cname);
  }

  protected function unregister() {
    return PGQ::unregister($this->log, $this->pg_src_con,
			   $this->qname, $this->cname);
  }

  protected function is_registered() {
    return PGQ::is_registered($this->log, $this->pg_src_con,
			      $this->qname, $this->cname);
  }

  protected function get_consumer_info() {
    return PGQ::get_consumer_info($this->log, $this->pg_src_con,
				  $this->qname, $this->cname);
  }

  public function get_consumers($log, $pgcon, $qname) {
    return PGQ::get_consumers($this->log, $this->pg_src_con, $this->qname);
  }

  protected function next_batch() {
    return PGQ::next_batch($this->log, $this->pg_src_con,
			   $this->qname, $this->cname);
  }

  protected function finish_batch($batch_id) {
    return PGQ::finish_batch($this->log, $this->pg_src_con, $batch_id);
  }

  protected function get_batch_events($batch_id) {
    return PGQ::get_batch_events($this->log, $this->pg_src_con, $batch_id);
  }

  protected function event_failed($batch_id, $event) {
    return PGQ::event_failed($this->log, $this->pg_src_con, $batch_id, $event);
  }

  protected function event_retry($batch_id, $event) {
    return PGQ::event_retry($this->log, $this->pg_src_con, $batch_id, $event);
  }

  protected function maint_retry_events() {
    return PGQ::maint_retry_events($this->log, $this->pg_src_con);
  }

  protected function failed_event_list() {
    return PGQ::failed_event_list($this->log, $this->pg_src_con,
				  $this->qname, $this->cname);
  }

  protected function failed_event_delete_all() {
    return PGQ::failed_event_delete_all($this->log, $this->pg_src_con,
					$this->qname, $this->cname);
  }

  protected function failed_event_delete($event_id) {
    return PGQ::failed_event_delete($this->log, $this->pg_src_con,
				    $this->qname, $this->cname, $event_id);
  }

  protected function failed_event_retry_all() {
    return PGQ::failed_event_retry_all($this->log, $this->pg_src_con,
				       $this->qname, $this->cname);
  }

  protected function failed_event_retry($event_id) {
    return PGQ::failed_event_retry($this->log, $this->pg_src_con,
				   $this->qname, $this->cname, $event_id);
  }
}
